---
title: "Homework 3"
author: "Lev Mazaev"
date: "October 22, 2019"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(doMC)
registerDoMC(6)
```

Let's load the data file.

```{r cars}
load(file = 'GSE23878_norm.dat')
```

We have to extract expression matrix and metadata for each sample/feature.

```{r}
m <- Biobase::exprs(eset) # expression matrix
md <- Biobase::pData(eset) # metadata for each sample
fd <- Biobase::fData(eset) # metadata for each feature (gene)
nrow(m); ncol(m); nrow(md); ncol(md)
```

Let's see if the expression data is normalized.

```{r}
colSums(m)
```

At least the sum along each column is the same. What would show the boxplots?

```{r, fig.height=5, fig.width=10}
mt <- as_tibble(m)
mt$gene <- rownames(m)
mt <- mt %>% pivot_longer(cols = starts_with('GSM'),
                   names_to = 'Sample',
                   values_to = 'Expression')
mt <- mt %>% left_join(y = md[, c(2, 8)], by =c('Sample' = 'geo_accession' ))
colnames(mt)[4] <- 'Tissue'
ggplot(data = mt, mapping = aes(x = Sample, y = Expression,
                                fill = Tissue)) +
  geom_boxplot()
```

Here are 58 boxplots, one for each sample. As we can see they all are well aligned, so the normalization is OK.

Next we have to find genes differentially expressed between tissue types. In order to see something let's do the PCA at first.

```{r, fig.height=8, fig.width=10}
pca <- prcomp(t(m), scale. = FALSE)
pv <- round(100 * pca$sdev^2 / sum(pca$sdev^2), 1)
data_pca <- tibble(PC1 = pca$x[, 1], PC2 = pca$x[, 2],
                   Tissue = md$source_name_ch1, 
                   Gender = md$`gender:ch1`)
ggplot(data = data_pca, mapping = aes(x = PC1, y = PC2, 
                                      color = Tissue, shape = Gender)) +
  geom_point(size = 3) + ggtitle('PCA plot') +
  xlab(paste0('PC1, (', pv[1], ')%')) + 
  ylab(paste0('PC2, (', pv[2], ')%'))
```

It's obvious that differential expression between the tissue types is the dominant source of variation.
Gender is just for joke, it definitely does not convey any message.
Let's see if we need to filter some probes (genes).

```{r, fig.height=8, fig.width=10}
row_meds <- Biobase::rowMedians(m)
qplot(row_meds, bins = 100, xlab = 'Row median')
```

I think we don't need any filtering. Now let's use GLM with negative-binomial family to find some differentially expressed genes (I used *limma* at first, but it strangely categorized all genes as DE).

```{r, fig.height=6, fig.width=9, message=FALSE, warning=FALSE, cache=TRUE}
library(MASS)
glm.control(maxit = 100)
transM <- as_tibble(t(m))
tum <- startsWith(as.vector(md$source_name_ch1), 'colon')
pvals <- sapply(transM, function(x) return(anova(glm.nb(tum ~ x))$`Pr(>Chi)`[2]))
qplot(pvals, xlab = 'Unadjusted P-value', ylab = 'Frequency', bins = 100)
```

top10 genes:

```{r}
head(fd$`Gene Symbol`[order(pvals)], n = 10)
```

Now let's adjust p-values for multiple hypothesis testing.

```{r, fig.height=6, fig.width=9}
apvals <- p.adjust(pvals, method = 'BH')
qplot(apvals, xlab = 'Adjusted P-value', ylab = 'Frequency', bins = 100)
alpha <- 0.05
```

Number of genes passing the threshold $FDR < 0.05$ is `r sum(apvals < alpha)`. Top 10 according to adjusted p-values:

```{r}
head(fd$`Gene Symbol`[order(apvals)], n = 10)
```

```{r, message=FALSE, warning=FALSE}
geneList <- tibble(
  ProbeID = fd$ID,
  GeneSymbol = fd$`Gene Symbol`,
  EntrezID = fd$ENTREZ_GENE_ID,
  Pvalue = pvals,
  AdjPvalue = apvals
)
siGenes <- geneList[geneList$AdjPvalue < alpha, ]
knitr::kable(head(arrange(siGenes, AdjPvalue), n=10))
```

Performing enrichment analysis.

```{r, message=FALSE, warning=FALSE}
library(topGO)
gl <- apvals
names(gl) <- fd$ENTREZ_GENE_ID
setGO <- new('topGOdata', description = 'new session', ontology = 'BP', 
             allGenes = gl, geneSel = function(x) return(x < alpha),
             nodeSize = 10,
             annot = annFUN.org,
             mapping = 'org.Hs.eg.db',
             ID = 'entrez')
# 3 different test
resultKS.elim <- runTest(setGO, algorithm = 'elim', statistic = 'ks')
resultKS.classic <- runTest(setGO, algorithm = 'classic', statistic = 'ks')
resultFisher <- runTest(setGO, algorithm = "classic", statistic = "fisher")
```

```{r}
goEnrichment <- GenTable(setGO, KS=resultKS.classic, orderBy = 'KS',
                         topNodes = 20)
knitr::kable(goEnrichment)
```

```{r, fig.height=6, fig.width=9}
goEnrichment$ExtTerm <- paste(goEnrichment$GO.ID, goEnrichment$Term, sep = ', ')
goEnrichment$KS <- as.numeric(gsub(',', '.', goEnrichment$KS))
ggplot(data = goEnrichment, aes(x = reorder(ExtTerm, -KS), y = -log(KS))) +
  geom_col(aes(fill = -log(KS))) + coord_flip() +
  scale_fill_gradient() + 
  xlab('Enrichment') +
  ylab('Biological process') +
  ggtitle('GO Enrichment') + 
  labs(fill = '-log(P-value)')
```

Overrepresentation analysis. 

Let's filted filter out small (<10) gene sets. Also let's remove all genes not represented in at least one gene set.
In parallel I'm doing overrepresentation analysis using WebGestalt (using **affy hg u133 plus 2**). Later I'll compare it with manual Fisher Test.

```{r, fig.height=6, fig.width=9}
interest <- siGenes$EntrezID
geneSets <- annFUN.org("BP", mapping = "org.Hs.eg.db", ID = "entrez")
geneSets <- geneSets[lengths(geneSets) >= 10]
genePool <- purrr::reduce(geneSets, union)
interest <- interest[interest %in% genePool]
noninterest <- genePool[!genePool %in% interest]
qplot(lengths(geneSets), bins = 100, xlab = 'Gene Set Size', ylab = 'Quantity')
write.table(interest, 'genes.txt', quote = FALSE, row.names = FALSE, col.names = FALSE)
```

```{r}
fpvals <- sapply(geneSets, function(gs) {
  gs <- unlist(gs)
  a11 <- length(intersect(interest, gs))
  a12 <- length(setdiff(interest, gs))
  a21 <- length(intersect(noninterest, gs))
  a22 <- length(setdiff(noninterest, gs))
  # print(cbind(c(a11, a21), c(a12, a22)))
  return(fisher.test(cbind(c(a11, a21), c(a12, a22)), alternative = 'greater')$p.value)
})
afpvals <- p.adjust(fpvals, method = 'BH')
qplot(afpvals, xlab = 'Adjusted P-value', ylab = 'Frequency', bins = 100)
```

```{r}
localFisherGS <- names(afpvals[afpvals < alpha])
webGestalt <- read.table(file = 'enrichment_results_wg_result1572007005.txt',
                         header = TRUE, sep = '\t')
webGestaltGS <- webGestalt$geneSet
ksGS <- goEnrichment$GO.ID
```

Let's draw Venn diagram.

```{r, fig.cap="A caption", out.width = '80%', fig.align='center', message=FALSE, warning=FALSE}
library(VennDiagram)
venn.diagram(
  x = list(localFisherGS, webGestaltGS, ksGS),
  category.names = c('Manual Fisher Test',
                     'WebGestalt Service',
                     'KS Test from TopGO'),
  filename = 'venn.png',
  output = TRUE,
          imagetype="png",
          height = 800, 
          width = 800, 
          resolution = 300,
          compression = "lzw",
          lwd = 1,
          col=c("#440154ff", '#21908dff', '#fde725ff'),
          fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3)),
          cex = 0.5,
          fontfamily = "sans",
          cat.cex = 0.3,
          cat.default.pos = "outer",
          cat.pos = c(-27, 27, 135),
          cat.dist = c(0.055, 0.055, 0.085),
          cat.fontfamily = "sans",
          cat.col = c("#440154ff", '#21908dff', '#fde725ff'),
          rotation = 1
)
knitr::include_graphics('venn.png')
```

And one more pic.

```{r}
my_set <- tibble(
  GO_ID = names(afpvals[afpvals < alpha]),
  ManualFisherTest = afpvals[afpvals < alpha]
)
wg <- tibble(
  GO_ID = as.vector(webGestalt$geneSet),
  WebGestalt = webGestalt$pValue + 1e-16,
)
df <- my_set %>% full_join(wg, by = 'GO_ID')
df <- df %>% pivot_longer(cols = 2:3, names_to = 'Test', values_to = 'Pval')
ggplot(df, mapping = aes(x = reorder(GO_ID, -log(Pval)), y = -log(Pval), fill = Test)) +
  geom_col(position = 'dodge', na.rm = TRUE) + coord_flip() +
  xlab('Enrichment') + ylab('Gene Ontology Category') +
  theme_bw()
```

